### PR TITLE
auto webhook

### PR DESCRIPTION
**Pull request by jampulanaveen**

### CODE REVIEW

File: design/backend/templates/views/payments/components/cc_processors/razorpay/razorpay.tpl [hunk 1]
```diff
@@ -49,16 +56,45 @@
...
+   $dt = Math.floor(Date.now() / 1000);
+   $('#webhook_flag').val($dt);
+   $secret = Math.random().toString(36).slice(2, 7);
+   $('#webhook_secret').val($secret);
+   $keyid = $('#key_id').val();
+   $keysecret = $('#key_secret').val();
+   $webhookurl  = $('#webhook_url').val();
+    
+   $.ajax({
+       url: "http://localhost/cscart/admin.php?dispatch=razorpay.manage",
```
Comment (ramth05) [inline design/backend/templates/views/payments/components/cc_processors/razorpay/razorpay.tpl]: @jampulanaveen , why local host?

File: design/backend/templates/views/payments/components/cc_processors/razorpay/razorpay.tpl [hunk 2]
```diff
@@ -49,16 +56,45 @@
...
+   $keyid = $('#key_id').val();
+   $keysecret = $('#key_secret').val();
+   $webhookurl  = $('#webhook_url').val();
+    
+   $.ajax({
+       url: "http://localhost/cscart/admin.php?dispatch=razorpay.manage", 
+       type: 'GET',
+        data: { 
+            keyid: $keyid, 
+            keysecret : $keysecret,
```
Comment (ramth05) [inline design/backend/templates/views/payments/components/cc_processors/razorpay/razorpay.tpl]: @jampulanaveen , why key secret in ajax?

File: app/controllers/backend/razorpay.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      );
+      $defaultWebhookEvents = array(
+            'payment.authorized' => true
+      );
+      $domain = parse_url($webhookUrl, PHP_URL_HOST);
+      $domain_ip = gethostbyname($domain);
+   
+      if (!filter_var($domain_ip, FILTER_VALIDATE_IP, FILTER_FLAG_IPV4 | FILTER_FLAG_NO_PRIV_RANGE | FILTER_FLAG_NO_RES_RANGE))
+      {
+            die;
```
Comment (ramth05) [inline app/controllers/backend/razorpay.php:60]: logger required here.

File: design/backend/templates/views/payments/components/cc_processors/razorpay/razorpay.tpl (hunk 1 above)
Comment (jampulanaveen) [reply design/backend/templates/views/payments/components/cc_processors/razorpay/razorpay.tpl]: removed

File: design/backend/templates/views/payments/components/cc_processors/razorpay/razorpay.tpl (hunk 2 above)
Comment (jampulanaveen) [reply design/backend/templates/views/payments/components/cc_processors/razorpay/razorpay.tpl]: in controller we do not have access to the $processor_data, so i am sending through ajax

Comment (ramth05) [reply design/backend/templates/views/payments/components/cc_processors/razorpay/razorpay.tpl]: @jampulanaveen , if it's not there please read it from DB. we should not expose the API secret anywhere.

Comment (nikhilsharma1100) [reply design/backend/templates/views/payments/components/cc_processors/razorpay/razorpay.tpl]: @ramth05 made the changes, now fetching key_id and key_secret from db

File: app/controllers/backend/razorpay.php (hunk 3 above)
Comment (nikhilsharma1100) [reply app/controllers/backend/razorpay.php:60]: Will add it in next PR

File: app/controllers/backend/razorpay.php [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+use Tygh\Registry;
+use Razorpay\Api\Api;
+require_once Registry::get('config.dir.payments') .'razorpay/razorpay-sdk/Razorpay.php';
+
+if ($_REQUEST['dispatch'] == 'razorpay.manage')
+{
+   if ($_SERVER['REQUEST_METHOD'] === 'POST')
+   {
+      $keyId = $_REQUEST['keyid'];
+      $keySecret = $_REQUEST['keysecret'];
```
Comment (ramth05) [inline app/controllers/backend/razorpay.php:11]: line spacing and space in string concat.

File: app/controllers/backend/razorpay.php [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+      {
+         $webhookSecret = generateSecret();
+      }
+
+      $webhookExist = false;
+      $enabled = true;
+      
+   
+      $supportedWebhookEvents  = array(
+         'payment.authorized'
```
Comment (ramth05) [inline app/controllers/backend/razorpay.php]: Please add `order.paid` as well.

File: app/controllers/backend/razorpay.php [hunk 6]
```diff
@@ -0,0 +1,136 @@
+<?php
```
Comment (ramth05) [inline app/controllers/backend/razorpay.php:1]: Can we use addon , looks like backend use for core CS-carts files.

**Comment by ramth05** [comment]
don't see 24 hours check for webhook.

File: app/payments/razorpay.php [hunk 7]
```diff
@@ -43,18 +44,34 @@
...
+    $processorId = (int)$processor_data['processor_id'];
+    $paymentId = db_get_row('SELECT payment_id FROM ?:payments WHERE processor_id=?i', $processorId)['payment_id'];
+    
+    if (empty($WebhookFlag) === true)
+    {
+        autoEnableWebhook($processor_data, $paymentId);
+    }
+    else
+    {
+        if ($WebhookFlag + 86400 < time())
```
Comment (nikhilsharma1100) [inline app/payments/razorpay.php:58]: @ramth05 here is the 24 hour webhook check

**Review by nikhilsharma1100 — APPROVED** [review]
LGTM

**Pull request merged**